src/rust/cryptography-x509-validation/src/policy/extension.rs

```diff
@@ -0,0 +1,486 @@
+        validator: Option<PresentExtensionValidatorCallback<B>>,
+    ) -> Self {
+        Self {
+            oid,
+            validator: ExtensionValidator::Present {
+                criticality,
+                validator,
+            },
+        }
+    }
+
+    pub(crate) fn maybe_present(
+        oid: ObjectIdentifier,
+        criticality: Criticality,
+        validator: Option<MaybeExtensionValidatorCallback<B>>,
+    ) -> Self {
+        Self {
+            oid,
+            validator: ExtensionValidator::MaybePresent {
+                criticality,
+                validator,
+            },
+        }
+    }
+
+    pub(crate) fn permits(
+        &self,
+        policy: &Policy<'_, B>,
+        cert: &Certificate<'_>,
+        extensions: &Extensions<'_>,
+    ) -> Result<(), PolicyError> {
+        match (&self.validator, extensions.get_extension(&self.oid)) {
+            // Extension MUST NOT be present and isn't; OK.
+            (ExtensionValidator::NotPresent, None) => Ok(()),
+            // Extension MUST NOT be present but is; NOT OK.
+            (ExtensionValidator::NotPresent, Some(_)) => Err(PolicyError::Other(
+                "EE certificate contains prohibited extension",
+            )),
+            // Extension MUST be present but is not; NOT OK.
+            (ExtensionValidator::Present { .. }, None) => Err(PolicyError::Other(
+                "EE certificate is missing required extension",
+            )),
+            // Extension MUST be present and is; check it.
+            (
+                ExtensionValidator::Present {
+                    criticality,
+                    validator,
+                },
+                Some(extn),
+            ) => {
+                if !criticality.permits(extn.critical) {
+                    return Err(PolicyError::Other(
+                        "EE certificate extension has incorrect criticality",
+                    ));
+                }
+
+                // If a custom validator is supplied, apply it.
+                validator.map_or(Ok(()), |v| v(policy, cert, &extn))
+            }
+            // Extension MAY be present.
+            (
+                ExtensionValidator::MaybePresent {
+                    criticality,
+                    validator,
+                },
+                extn,
+            ) => {
+                // If the extension is present, apply our criticality check.
+                if extn
+                    .as_ref()
+                    .map_or(false, |extn| !criticality.permits(extn.critical))
+                {
+                    return Err(PolicyError::Other(
+                        "EE certificate extension has incorrect criticality",
+                    ));
+                }
+
+                // If a custom validator is supplied, apply it.
+                validator.map_or(Ok(()), |v| v(policy, cert, extn.as_ref()))
+            }
+        }
+    }
+}
+
+#[allow(dead_code)]
+pub(crate) mod ee {
```

Comment:
Per non-review comment: I think we should probably do the `ee`, `ca`, and `common` submodules here as a subsequent breakout to keep the PR small.

(Alternatively, we could just leave those for #8873, since we'll get coverage on them for free via Limbo.)

Reply:
Yeah, that'd be ideal actually! Unit-testing each of the validators was going to be pretty laborious. I'll remove the submodules and get this in first.